librad2/src/keys/pgp.rs

```diff
@@ -0,0 +1,91 @@
+use std::fmt;
+use std::time::SystemTime;
+
+use pbp;
+use sha2::{Digest, Sha256};
+
+pub trait Signer {
+    fn sign(&self, data: &[u8]) -> [u8; 64];
+}
+
+pub trait Verifier {
+    fn verify(&self, data: &[u8], signature: [u8; 64]) -> bool;
+}
+
+pub struct Key<'a, S: Signer> {
+    signer: &'a S,
+    key: pbp::PgpKey,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Signature(pbp::PgpSig);
+
+pub type Error = pbp::PgpError;
+
+impl<'a, S: Signer> Key<'a, S> {
+    pub fn new(key: &[u8], user_id: &str, unix_time: u32, signer: &'a S) -> Self {
+        let key =
+            pbp::PgpKey::new::<Sha256, _>(key, pbp::KeyFlags::SIGN, user_id, unix_time, |data| {
+                signer.sign(data)
+            });
+        Self { signer, key }
+    }
+
+    pub fn sign(&self, data: &[u8]) -> Signature {
+        let timestamp = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap()
+            .as_secs();
+
+        Signature(pbp::PgpSig::new::<Sha256, _>(
+            data,
+            self.key.fingerprint(),
+            pbp::SigType::Standalone,
+            timestamp as u32,
+            &[],
+            |data| self.signer.sign(data),
+        ))
+    }
+}
+
+impl<S: Signer> fmt::Display for Key<'_, S> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.key.fmt(f)
+    }
+}
+
+impl Signature {
+    pub fn verify<F: Verifier>(&self, data: &[u8], verifier: F) -> bool {
+        self.0.verify::<Sha256, _, _>(
+            |hasher| hasher.input(data),
+            |hashed, sig| verifier.verify(hashed, sig),
+        )
+    }
+
+    pub fn from_ascii_armor(ascii: &str) -> Result<Signature, Error> {
+        pbp::PgpSig::from_ascii_armor(ascii).map(Signature)
+    }
+}
+
+impl fmt::Display for Signature {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use super::*;
+
+    #[test]
+    #[ignore]
+    fn bananas() -> Result<(), Error> {
```

Comment:
@FintanH It's very dissapointing: this nifty `pbp` create _sometimes_ produces ASCII armor it can't itself read back in. Knee-deep into `sequoia-openpgp` again....

Reply:
![image](https://user-images.githubusercontent.com/3503528/69422101-c2fea300-0d1a-11ea-9869-bc434cdab7bb.png)